libcontainer/system/proc.go

```diff
@@ -75,29 +78,42 @@ func parseStat(data string) (stat Stat_t, err error) {
 	// From proc(5), field 2 could contain space and is inside `(` and `)`.
 	// The following is an example:
 	// 89653 (gunicorn: maste) S 89630 89653 89653 0 -1 4194560 29689 28896 0 3 146 32 76 19 20 0 1 0 2971844 52965376 3920 18446744073709551615 1 1 0 0 0 0 0 16781312 137447943 0 0 0 17 1 0 0 0 0 0 0 0 0 0 0 0 0 0
-	i := strings.LastIndex(data, ")")
+	i := strings.LastIndexByte(data, ')')
 	if i <= 2 || i >= len(data)-1 {
-		return stat, fmt.Errorf("invalid stat data: %q", data)
+		return stat, fmt.Errorf("invalid stat data (no comm): %q", data)
 	}
 
-	parts := strings.SplitN(data[:i], "(", 2)
+	parts := strings.SplitN(data[:i], " (", 2)
 	if len(parts) != 2 {
-		return stat, fmt.Errorf("invalid stat data: %q", data)
+		return stat, fmt.Errorf("invalid stat data (no comm): %q", data)
 	}
 
 	stat.Name = parts[1]
-	_, err = fmt.Sscanf(parts[0], "%d", &stat.PID)
+	pid, err := strconv.ParseUint(parts[0], 10, bits.UintSize)
 	if err != nil {
-		return stat, err
+		return stat, errors.Wrap(err, "invalid stat data (bad pid)")
+	}
+	stat.PID = uint(pid)
+
+	// Remove the first two fields and a space after.
+	data = data[i+2:]
+	stat.State = State(data[0])
+
+	// StartTime is field 22, data is at field 3 now, so we need to skip 19 spaces.
+	skipSpaces := 22 - 3
+	for i = 0; skipSpaces > 0 && i < len(data); i++ {
+		if data[i] == ' ' {
+			skipSpaces--
+		}
+	}
+	j := strings.IndexByte(data[i:], ' ')
+	if j <= 0 {
+		return stat, fmt.Errorf("invalid stat data (too short): %q", data)
```

Comment:
```suggestion
		return stat, errors.Errorf("invalid stat data (too short): %q", data)
```

Reply:
So the differences are
 - errors package is not standard
 - it encapsulates the calling stack (i.e. backtrace) into the error (which might be very handy, but I've never seen it used anywhere)
 - anything else that I miss?
 
As this code does not have many callers, I think the backtrace does not make much sense, does it?

Reply:
Backtrace is used for `--debug` I think? If it isn't, it should be (that's what I use in umoci).

Reply:
We did not do that, but it's trivial to fix: https://github.com/opencontainers/runc/pull/2789

Reply:
fixed to use pkg/errors